Implement `Index` and `IndexMut` traits for `ArrayVec`

niluxv opened the pull request:
I didn't have time to add tests to it yet, but this should implement `Index<I>` and `IndexMut<I>` where `I: SliceIndex<[T]>`.

Fixes #190

niluxv committed:
Implement `Index` and `IndexMut` traits for `ArrayVec`

niluxv commented:
Instead of this implementation (`self.get[_mut](index).unwrap()`) we could also use
```rust
self.deref().index[_mut](index)
```
Maybe that gives better error messages in some situations? I don't know...

bluss commented:
Looks to be in the right direction, that's nice - I'll be a bit unreachable during the summer, but back in a bit

niluxv committed:
Add tests for `Index` and `IndexMut` trait implements

niluxv commented:
No problem.

I changed the implementation as described above because it indeed gives much better error messages on out of bounds errors, and added some tests.

bluss commented on src/arrayvec.rs:
```diff
@@ -1002,6 +1002,42 @@ impl<T, const CAP: usize> Extend<T> for ArrayVec<T, CAP> {
+    /// # Panics
+    /// Panics if index is out of bounds.
+    fn index(&self, index: I) -> &Self::Output {
+        self.deref().index(index)
```
Now I guess this is a matter of taste, but IMO we don't call operator methods explicitly, we just use the relevant operator. The operator for deref is `*` and index is, you know, etc.. :slightly_smiling_face: 

```suggestion
        (**self)[index]
```

niluxv replied:
Usually I also prefer the syntactic sugar, but when implementing `index` I always use `index` explicitly. Mainly a matter of preference. Changed it now. (The generated mir seems nearly identical. The new version uses only 1 additional local and assignment, so I don't expect a difference in runtime performance.)

bluss commented on src/arrayvec.rs:
```diff
@@ -1002,6 +1002,42 @@ impl<T, const CAP: usize> Extend<T> for ArrayVec<T, CAP> {
+    /// # Panics
+    /// Panics if index is out of bounds.
+    fn index_mut(&mut self, index: I) -> &mut Self::Output {
+        self.deref_mut().index_mut(index)
```
```suggestion
        (**self)[index]
```

bluss reviewed (changes requested):
Thanks, yes not regressing the panic messages is certainly a requirement. One could also debate if `I: SliceIndex` or if `[T]: Index<I>` is the trait bound we should use (the latter being yet more general), but it seems fine as it is now.

We need the corresponding change for ArrayString too, but it can be a different PR.

niluxv committed:
Change notation in Index[Mut] implementation

niluxv committed:
Make trait bound on Index[Mut] implementation more general

niluxv committed:
Implement `Index` and `IndexMut` traits for `ArrayString`

niluxv commented:
So I changed the trait bounds as you suggested and just now added implementations to `ArrayString` analogous to those for `ArrayVec`.
I also removed the documentation I initially added to the `Index`/`IndexMut` implementation since the comment on `panic`ing needn't be true for all instantiations of the generic parameter.

mikwielgus commented:
Hello, may I ask what is the status of this PR? I see that this was *approved* but not merged, is there anything stalling this beyond merge conflict resolution? It would be convenient for me to have `Index` implemented for `ArrayVec`.
